### ISSUE TITLE
Submitting an AWS job with Python 3

### ISSUE THREAD

**Issue by TMiguelT** [DataBiosphere/toil#2736]
I'm on Python 3.6, Toil 3.19.0.

I've successfully set up an AWS toil cluster using the tutorial provided, but if I try to run a job, I get:

```bash
$ python workflow.py aws:ap-southeast-2:toil
Traceback (most recent call last):
  File "workflow.py", line 90, in <module>
    with Toil(options) as toil:
  File "/home/michael/Programming/1000gMultiQc/venv/lib/python3.6/site-packages/toil/common.py", line 699, in __enter__
    jobStore.initialize(config)
  File "/home/michael/Programming/1000gMultiQc/venv/lib/python3.6/site-packages/toil/jobStores/aws/jobStore.py", line 137, in initialize
    super(AWSJobStore, self).initialize(config)
  File "/home/michael/Programming/1000gMultiQc/venv/lib/python3.6/site-packages/toil/jobStores/abstractJobStore.py", line 137, in initialize
    self.writeConfig()
  File "/home/michael/Programming/1000gMultiQc/venv/lib/python3.6/site-packages/toil/jobStores/abstractJobStore.py", line 144, in writeConfig
    with self.writeSharedFileStream('config.pickle', isProtected=False) as fileHandle:
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/home/michael/Programming/1000gMultiQc/venv/lib/python3.6/site-packages/toil/jobStores/aws/jobStore.py", line 537, in writeSharedFileStream
    encrypted=isProtected)
  File "/home/michael/Programming/1000gMultiQc/venv/lib/python3.6/site-packages/toil/jobStores/aws/jobStore.py", line 901, in loadOrCreate
    self = cls.load(jobStoreFileID)
  File "/home/michael/Programming/1000gMultiQc/venv/lib/python3.6/site-packages/toil/jobStores/aws/jobStore.py", line 895, in load
    cls.outer.filesDomain.get_attributes(item_name=bytes(jobStoreFileID),
TypeError: string argument without an encoding
```

Any idea why this is happening?

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-395)
┆Issue Number: TOIL-395

**Comment by TMiguelT** [comment]
I had a look into this, and `bytes("some string")` only ever worked in Python 2, and never in Python 3. So this seems like a fundamental issue in the job store code. Should I PR?

**Comment by mr-c** [comment]
@TMiguelT Yes, please! Bonus if you can derive a unit test to prove the fix

**Comment by TMiguelT** [comment]
I'll look into it. Are we still ensuring Python 2 compatibility as well?

**Comment by TMiguelT** [comment]
A look through the codebase finds about 40 of these invocations. Has no-one ever used Python 3 with a cluster before? 

```
./src/toil/jobStores/aws/jobStore.py:                    item_name=bytes(jobStoreID),
./src/toil/jobStores/aws/jobStore.py:                item = self.jobsDomain.get_attributes(bytes(jobStoreID), consistent_read=True)
./src/toil/jobStores/aws/jobStore.py:                assert self.jobsDomain.put_attributes(bytes(job.jobStoreID), item)
./src/toil/jobStores/aws/jobStore.py:                item = self.jobsDomain.get_attributes(bytes(jobStoreID), consistent_read=True)
./src/toil/jobStores/aws/jobStore.py:                self.jobsDomain.delete_attributes(item_name=bytes(jobStoreID))
./src/toil/jobStores/aws/jobStore.py:                            self.filesBucket.delete_key(key_name=bytes(item.name), version_id=version)
./src/toil/jobStores/aws/jobStore.py:                            self.filesBucket.delete_key(key_name=bytes(item.name))
./src/toil/jobStores/aws/jobStore.py:                key = self.filesBucket.get_key(key_name=bytes(jobStoreFileID), version_id=info.version)
./src/toil/jobStores/aws/jobStore.py:                        item_name=bytes(jobStoreFileID),
./src/toil/jobStores/aws/jobStore.py:                        cls.outer.filesDomain.get_attributes(item_name=bytes(jobStoreFileID),
./src/toil/jobStores/aws/jobStore.py:                        assert self.outer.filesDomain.put_attributes(item_name=bytes(self.fileID),
./src/toil/jobStores/aws/jobStore.py:                            self.outer.filesBucket.delete_key(bytes(self.fileID),
./src/toil/jobStores/aws/jobStore.py:                            self.outer.filesDomain.delete_attributes(bytes(self.fileID),
./src/toil/jobStores/aws/jobStore.py:                                              bucket=self.outer.filesBucket, fileID=bytes(self.fileID),
./src/toil/jobStores/aws/jobStore.py:                                    key_name=bytes(info.fileID),
./src/toil/jobStores/aws/jobStore.py:                        key = store.filesBucket.new_key(key_name=bytes(info.fileID))
./src/toil/jobStores/aws/jobStore.py:                        srcKey = self.outer.filesBucket.get_key(bytes(self.fileID), headers=self._s3EncryptionHeaders())
./src/toil/jobStores/aws/jobStore.py:                        srcKey = self.outer.filesBucket.get_key(bytes(self.fileID))
./src/toil/jobStores/aws/jobStore.py:                key = self.outer.filesBucket.get_key(bytes(self.fileID), validate=False)
./src/toil/jobStores/aws/jobStore.py:                        key = info.outer.filesBucket.get_key(bytes(info.fileID), validate=False)
./src/toil/jobStores/aws/jobStore.py:                            bytes(self.fileID),
./src/toil/jobStores/aws/jobStore.py:                            store.filesBucket.delete_key(key_name=bytes(self.fileID),
./src/toil/jobStores/aws/utils.py:        key = bucket.new_key(key_name=bytes(fileID))
./src/toil/jobStores/aws/utils.py:            key = bucket.get_key(bytes(fileID),
./src/toil/jobStores/aws/utils.py:                key_name=bytes(fileID),
./src/toil/jobStores/googleJobStore.py:                self.sseKey = bytes(f.read())
./src/toil/jobStores/googleJobStore.py:        return self.bucket.blob(bytes(jobStoreID), encryption_key=self.sseKey).exists()
./src/toil/jobStores/googleJobStore.py:        blob = self.bucket.get_blob(bytes(fileName), encryption_key=self.sseKey)
./src/toil/jobStores/googleJobStore.py:        for blob in self.bucket.list_blobs(prefix=bytes(jobStoreID)):
./src/toil/jobStores/googleJobStore.py:            blob = self.bucket.get_blob(bytes(jobStoreFileID), encryption_key=self.sseKey)
./src/toil/jobStores/googleJobStore.py:        return self.bucket.blob(bytes(jobStoreFileID), encryption_key=self.sseKey).exists()
./src/toil/jobStores/googleJobStore.py:        blob = bucket.blob(bytes(fileName))
./src/toil/jobStores/googleJobStore.py:            for blob in self.bucket.list_blobs(prefix=bytes(prefix)):
./src/toil/jobStores/googleJobStore.py:                        self.bucket.rename_blob(blob, bytes(newID))
./src/toil/jobStores/googleJobStore.py:            self.bucket.get_blob(bytes(jobStoreFileID)).delete()
./src/toil/jobStores/googleJobStore.py:        job = self.bucket.get_blob(bytes(jobStoreID), encryption_key=self.sseKey)
./src/toil/jobStores/googleJobStore.py:        blob = self.bucket.blob(bytes(jobStoreID), encryption_key=self.sseKey if encrypt else None)
./src/toil/jobStores/googleJobStore.py:        blob = self.bucket.blob(bytes(fileName), encryption_key=self.sseKey if encrypt else None)
./src/toil/jobStores/googleJobStore.py:        blob = self.bucket.get_blob(bytes(fileName), encryption_key=self.sseKey if encrypt else None)
```

**Comment by TMiguelT** [comment]
We definitely need an offline (travis-able) test for these cloud job stores as well, to pick these errors up. But I'm not certain how to go about writing one.

**Comment by adamnovak** [comment]
We might not use Python 3 at all in cluster mode. I think `toil
launch-cluster` deploys Python 2, and our cloud integration tests
aren't up and running yet.

**Comment by TMiguelT** [comment]
Right. That might be something to reconsider with the Python 2 EOL. But in any case, it's the code that runs locally when you submit a job to the cluster that fails because of Python 3, not the code that sets up the cluster.

**Comment by adamnovak** [comment]
If you used `toil launch-cluster` to create a cluster, the right way to actually use it is `toil ssh-cluster -z ap-southeast-2 clustername`, and then to run your `workflow.py` on the head node machine you end up on in that SSH session, with `--batchSystem mesos` to tell it to use the Mesos instance that launch-cluster made, in order to send jobs to the cluster nodes. If you didn't provision any cluster nodes when you launched the cluster, you need a couple more arguments to tell it to auto-scale the cluster. See the last command [here in this part of the tutorial](https://toil.readthedocs.io/en/latest/running/cloud/amazon.html#running-a-workflow-with-autoscaling):

```
$ python /root/sort.py aws:us-west-2:<my-jobstore-name> --provisioner aws --nodeTypes c3.large --maxNodes 2 --batchSystem mesos
```

Note that we/re running out of /root, because on the Toil-deployed head node, you get to be root. Within the head node, `python` is currently Python 2 by default, which is how we avoid running into this error you've reported most of the time. We do need to make it work with Python 3, but running your workflow in Python 3 **on your local machine** is not going to use the cluster you launched, even when we do get it working.

Looking again at your `python workflow.py aws:ap-southeast-2:toil` command, and where you've run it, it will just run the whole workflow on your local machine (because the default batch system is single machine), but using AWS S3 to store the job data and workflow intermediate state.

**Comment by TMiguelT** [comment]
You're right, I was executing the workflow incorrectly. However I've since used `toil ssh-cluster` to access the head node and attempted to submit a Python 3 job, but I encounter countless errors. An awful lot of the AWS code is not compatible with Python 3.

I've started a Python 3 fix Pull Request, which is here: #2737. However, after fixing these minor syntax issues I've hit a problem I can't fix, namely that the Python 2 worker nodes are expecting a different version of the pickle'd data:
```
ip-172-31-3-200.ap-southeast-2.compute.internal 2019-07-18 07:42:29,482 Process IO WARNING toil.batchSystems.mesos.batchSystem: Job 0 failed with message 'ValueError: unsupported pickle protocol: 3
' due to reason '{}' on executor '{'value': 'toil-1691'}' on agent '{'value': '3849f37a-b9b2-4f75-8135-ec3c9fd3f5ea-S2'}'.
```

I'm also still getting some more complex bytes/string mismatches:
```
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/root/venv3/lib/python3.5/site-packages/toil/statsAndLogging.py", line 156, in statsAndLoggingAggregator
    jobStore.writeStatsAndLogging(text)
  File "/root/venv3/lib/python3.5/site-packages/toil/jobStores/aws/jobStore.py", line 592, in writeStatsAndLogging
    writeable.write(statsAndLoggingString)
TypeError: a bytes-like object is required, not 'str'
```

I had hoped to quickly fix these issues, but it seems that the whole codebase is heavily tied to Python 2, and I sadly don't have time to fix that. Anyway I'm leaving my half-done PR and this issue for any future work that might happen.

**Comment by adamnovak** [comment]
The cluster we build with `toil launch-cluster` is probably pretty tied to Python 2 because that's what the default is in the Docker images used to set up the cluster. If you want to run Python 3 workflows, you will probably need a custom `TOIL_APPLIANCE_SELF` Docker image that has Toil installed on Python 3, in addition to squashing all the Python 3 compatibility bugs.
